crates/resolver/src/name_server/name_server_stats.rs

```diff
@@ -6,125 +6,295 @@
+/// the returned value decreases.
+fn compute_srtt_factor(last_update: Instant, weight: u32) -> f64 {
+   let exponent = (-last_update.elapsed().as_secs_f64().max(1.0)) / f64::from(weight);
+   exponent.exp()
 }
 
 impl NameServerStats {
-    pub(crate) fn new(successes: usize, failures: usize) -> Self {
-        Self {
-            successes: AtomicUsize::new(successes),
-            failures: AtomicUsize::new(failures),
-        }
-    }
-
-    pub(crate) fn next_success(&self) {
-        self.successes.fetch_add(1, atomic::Ordering::Release);
-    }
-
-    pub(crate) fn next_failure(&self) {
-        self.failures.fetch_add(1, atomic::Ordering::Release);
-    }
-
-    fn noload_eq(
-        self_successes: usize,
-        other_successes: usize,
-        self_failures: usize,
-        other_failures: usize,
-    ) -> bool {
-        self_successes == other_successes && self_failures == other_failures
-    }
+   const CONNECTION_FAILURE_PENALTY: u32 = Duration::from_millis(150).as_micros() as u32;
+   const MAX_SRTT_MICROS: u32 = Duration::from_secs(5).as_micros() as u32;
+
+   pub(crate) fn new(initial_srtt: u32) -> Self {
+       Self {
+           srtt_microseconds: AtomicU32::new(initial_srtt),
+           last_update: Arc::new(Mutex::new(None)),
+       }
+   }
+
+   /// Records the measured `rtt` for a particular query.
+   pub(crate) fn record_rtt(&self, rtt: Duration) {
+       // If the cast on the result does overflow (it shouldn't), then the
+       // value is saturated to u32::MAX, which is above the `MAX_SRTT_MICROS`
+       // limit (meaning that any potential overflow is inconsequential).
+       // See https://github.com/rust-lang/rust/issues/10184.
+       self.update_srtt(rtt.as_micros() as u32, |cur_srtt_microseconds, last_update| {
+           // An arbitrarily low weight is used when computing the factor
+           // to ensure that recent RTT measurements are weighted more
+           // heavily.
+           let factor = compute_srtt_factor(last_update, 3);
+           let new_srtt = (1.0 - factor) * (rtt.as_micros() as f64) + factor * f64::from(cur_srtt_microseconds);
+           new_srtt.round() as u32
+       });
+   }
+
+   /// Records a connection failure for a particular query.
+   pub(crate) fn record_connection_failure(&self) {
+       self.update_srtt(Self::CONNECTION_FAILURE_PENALTY, |cur_srtt_microseconds, _last_update| {
+           cur_srtt_microseconds.saturating_add(Self::CONNECTION_FAILURE_PENALTY)
+       });
+   }
+
+   /// Returns the raw SRTT value.
+   ///
+   /// Prefer to use `decayed_srtt` when ordering name servers.
+   fn srtt_microseconds(&self) -> u32 {
+       self.srtt_microseconds.load(atomic::Ordering::Acquire)
+   }
+
+   /// Returns the SRTT value after applying a time based decay.
+   ///
+   /// Prefer to use this value when ordering name servers. Applying a decay
+   /// helps distribute query load and detect positive network changes.
+   fn decayed_srtt(&self) -> f64 {
```

Comment:
Would be nice to add more discussion here of how the value decays (that is, does it decay upwards or downwards) and why.

Reply:
Done.